src/parser.c

```diff
@@ -494,53 +584,68 @@ struct l_queue *parser_queue_from_json_array(json_object *jobj,
 
 	for (i = 0; i < len; i++) {
 		jobjentry = json_object_array_get_idx(jarray, i);
-
-		item = foreach_cb(jobjentry);
+		item = device_array_item(jobjentry, item_cb);
 		if (item)
 			l_queue_push_tail(list, item);
 	}
-
+	json_object_put(jobj);
 	return list;
 }
 
-struct l_queue *parser_request_to_list(json_object *jso)
+struct l_queue *parser_request_to_list(const char *json_str)
 {
+	json_object *jso;
 	struct l_queue *list;
 	json_object *json_array;
 	json_object *jobjentry;
-	int sensor_id;
+	int sensor_id;	
 	uint64_t i;
+	bool has_err;
 
+	jso = json_tokener_parse(json_str);
+	if (!jso)
+		return NULL;
 	list = l_queue_new();
 
-	if (!json_object_object_get_ex(jso, "sensorIds", &json_array))
-		goto fail;
+	if (!json_object_object_get_ex(jso, KNOT_JSON_FIELD_SENSOR_IDS,
+			&json_array)) {
+		l_queue_destroy(list, l_free);
+		return NULL;
+	}
 
+	has_err = false;			
 	for (i = 0; i < json_object_array_length(json_array); i++) {
 
 		jobjentry = json_object_array_get_idx(json_array, i);
-		if (!jobjentry)
-			goto fail;
+		if (!jobjentry) {
+			has_err = true;
+			break;
+		}
 
-		if (json_object_get_type(jobjentry) != json_type_int)
-			goto fail;
+		if (json_object_get_type(jobjentry) != json_type_int) {
+			has_err = true;
+			break;
+		}
 
 		sensor_id = json_object_get_int(jobjentry);
 
 		if (!l_queue_push_tail(list,
-				l_memdup(&sensor_id, sizeof(sensor_id))))
-			goto fail;
+				l_memdup(&sensor_id, sizeof(sensor_id)))) {
+			has_err = true;
+			break;
+		}
+	}
+	
+	if (has_err) {
+		l_queue_destroy(list, l_free);
+		return NULL;
 	}
-
```

Comment:
Is removing this line really necessary?

Reply:
I think it is, because the next statement is a return